File: indiescraper/src/robots.rs
use reqwest;
use std::collections::HashMap;
use std::error::Error;
use url::Url;

#[derive(Debug, Clone)]
pub struct RobotsRule {
    pub user_agent: String,
    pub allowed_paths: Vec<String>,
    pub disallowed_paths: Vec<String>,
}

#[derive(Debug)]
pub struct RobotsResult {
    pub allowed: Vec<String>,
    pub disallowed: Vec<String>,
}

pub async fn check_robots_txt(url: &str) -> Result<Option<RobotsResult>, Box<dyn Error>> {
    let base_url = Url::parse(url)?;
    let robots_url = format!("{}/robots.txt", base_url.origin().ascii_serialization());
    
    let client = reqwest::Client::builder()
        .timeout(std::time::Duration::from_secs(5))
        .build()?;
    
    let response = client.get(&robots_url).send().await?;
    
    if !response.status().is_success() {
        return Ok(None);
    }
    
    let content = response.text().await?;
    
    if content.is_empty() {
        return Ok(None);
    }
    
    if content.len() > 5000 {
        return Ok(None);
    }
    
    let rules = parse_robots_txt(&content);
    let allowed_paths = get_allowed_paths(&rules, &["*", "indieseas"], &base_url.origin().ascii_serialization());
    
    let wildcard = allowed_paths.get("*").cloned().unwrap_or_default();
    let indieseas = allowed_paths.get("indieseas").cloned().unwrap_or_default();
    
    Ok(Some(RobotsResult {
        allowed: [wildcard.allowed, indieseas.allowed].concat(),
        disallowed: [wildcard.disallowed, indieseas.disallowed].concat(),
    }))
}

fn parse_robots_txt(content: &str) -> Vec<RobotsRule> {
    let lines: Vec<&str> = content.lines().collect();
    let mut rules = Vec::new();
    let mut current_rule: Option<RobotsRule> = None;
    
    if lines.is_empty() {
        return rules;
    }
    
    for line in lines {
        let line = line.split('#').next().unwrap_or("").trim().to_lowercase();
        if line.is_empty() {
            continue;
        }
        
        if line.starts_with("user-agent:") {
            let user_agent = line.strip_prefix("user-agent:").unwrap_or("").trim().to_string();
            
            if let Some(ref mut rule) = current_rule {
                if rule.user_agent == user_agent {
                    continue;
                }
            }
            
            let new_rule = RobotsRule {
                user_agent,
                allowed_paths: Vec::new(),
                disallowed_paths: Vec::new(),
            };
            
            if let Some(rule) = current_rule.take() {
                rules.push(rule);
            }
            current_rule = Some(new_rule);
        } else if line.starts_with("allow:") {
            let path = line.strip_prefix("allow:").unwrap_or("").trim();
            if !path.is_empty() {
                if let Some(ref mut rule) = current_rule {
                    rule.allowed_paths.push(path.to_string());
                } else {
                    current_rule = Some(RobotsRule {
                        user_agent: "*".to_string(),
                        allowed_paths: vec![path.to_string()],
                        disallowed_paths: Vec::new(),
                    });
                }
            }
        } else if line.starts_with("disallow:") {
            let path = line.strip_prefix("disallow:").unwrap_or("").trim();
            if !path.is_empty() {
                if let Some(ref mut rule) = current_rule {
                    rule.disallowed_paths.push(path.to_string());
                } else {
                    current_rule = Some(RobotsRule {
                        user_agent: "*".to_string(),
                        allowed_paths: Vec::new(),
                        disallowed_paths: vec![path.to_string()],
                    });
                }
            }
        }
    }
    
    if let Some(rule) = current_rule {
        rules.push(rule);
    }
    
    rules
}

#[derive(Debug, Default, Clone)]
struct PathResult {
    allowed: Vec<String>,
    disallowed: Vec<String>,
}

fn get_allowed_paths(rules: &[RobotsRule], user_agents: &[&str], base_url: &str) -> HashMap<String, PathResult> {
    let mut result = HashMap::new();
    let normalized_user_agents: Vec<String> = user_agents.iter().map(|ua| ua.to_lowercase()).collect();
    
    // Initialize result with all requested user agents
    for ua in &normalized_user_agents {
        result.insert(ua.clone(), PathResult::default());
    }
    
    // Find wildcard rules first to use as fallback
    let wildcard_rules: Vec<&RobotsRule> = rules.iter().filter(|rule| rule.user_agent == "*").collect();
    
    // Find applicable rules for each user agent
    for ua in &normalized_user_agents {
        let specific_rules: Vec<&RobotsRule> = rules.iter().filter(|rule| rule.user_agent == *ua).collect();
        let applicable_rules = if !specific_rules.is_empty() { specific_rules } else { wildcard_rules.clone() };
        
        let mut path_result = PathResult::default();
        
        if applicable_rules.is_empty() {
            // No specific rules and no wildcard rules, so everything is allowed by default
            path_result.allowed.push(format!("{}/", base_url));
        } else {
            for rule in applicable_rules {
                // If no disallow rules, everything is allowed
                if rule.disallowed_paths.is_empty() {
                    path_result.allowed.push(format!("{}/", base_url));
                    continue;
                }
                
                // Add all explicitly allowed paths
                for path in &rule.allowed_paths {
                    let mut normalized_path = path.replace('*', "");
                    if !normalized_path.starts_with('/') {
                        normalized_path = format!("/{}", normalized_path);
                    }
                    path_result.allowed.push(format!("{}{}", base_url, normalized_path));
                }
                
                // Add all disallowed paths
                for path in &rule.disallowed_paths {
                    let mut normalized_path = path.replace('*', "");
                    if !normalized_path.starts_with('/') {
                        normalized_path = format!("/{}", normalized_path);
                    }
                    path_result.disallowed.push(format!("{}{}", base_url, normalized_path));
                }
            }
            
            // If no specific allowed paths but disallow paths exist, root path is allowed
            let root_path = format!("{}/", base_url);
            if path_result.allowed.is_empty() && !path_result.disallowed.contains(&root_path) {
                path_result.allowed.push(root_path);
            }
        }
        
        result.insert(ua.clone(), path_result);
    }
    
    result
}

pub async fn is_allowed(domain: &str) -> Result<bool, Box<dyn Error>> {
    let robots_url = if domain.ends_with('/') {
        format!("{}robots.txt", domain)
    } else {
        format!("{}/robots.txt", domain)
    };
    
    let client = reqwest::Client::builder()
        .timeout(std::time::Duration::from_secs(5))
        .build()?;
    let response = client.get(&robots_url).send().await?;
    let content = response.text().await?;
    
    let mut applies_to_indieseas = false;
    
    for line in content.lines() {
        let line = line.trim();
        
        if line.starts_with("User-agent:") {
            let agent = line.split(':').nth(1).unwrap_or("").trim();
            applies_to_indieseas = agent == "*" || agent.to_lowercase() == "indieseas";
        } else if line.starts_with("Disallow:") && applies_to_indieseas {
            let path = line.split(':').nth(1).unwrap_or("").trim();
            if path == "/" {
                return Ok(false);
            }
        } else if line.is_empty() {
            applies_to_indieseas = false;
        }
    }
    
    Ok(true)
}

File: scraper/robots.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"log"
	"net/http"
	"net/url"
	"strings"
	"time"
)

type RobotsRule struct {
	UserAgent      string
	AllowedPaths   []string
	DisallowedPaths []string
}

type RobotsResult struct {
	Allowed    []string
	Disallowed []string
}

func CheckRobotsTxt(rawURL string) (*RobotsResult, error) {
	parsed, err := url.Parse(rawURL)
	if err != nil {
		return nil, fmt.Errorf("invalid url %q: %w", rawURL, err)
	}
	base := fmt.Sprintf("%s://%s", parsed.Scheme, parsed.Host)
	robotsURL := base + "/robots.txt"

	client := &http.Client{Timeout: 5 * time.Second}
	resp, err := client.Get(robotsURL)
	if err != nil {
		log.Printf("couldn't fetch robots.txt: %v", err)
		return nil, nil
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, nil
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("reading robots.txt: %w", err)
	}
	if len(body) == 0 {
		return nil, nil
	}
	if len(body) > 5000 {
		return nil, nil
	}

	rules := parseRobotsTxt(string(body))
	paths := getAllowedPaths(rules, []string{"*", "indieseas"}, base)

	allowed := append(paths["*"].Allowed, paths["indieseas"].Allowed...)
	disallowed := append(paths["*"].Disallowed, paths["indieseas"].Disallowed...)

	return &RobotsResult{Allowed: allowed, Disallowed: disallowed}, nil
}

func parseRobotsTxt(content string) []RobotsRule {
	scanner := bufio.NewScanner(strings.NewReader(content))
	var rules []RobotsRule
	var current *RobotsRule

	for scanner.Scan() {
		line := scanner.Text()
		// strip comments and whitespace
		if idx := strings.Index(line, "#"); idx >= 0 {
			line = line[:idx]
		}
		line = strings.TrimSpace(strings.ToLower(line))
		if line == "" {
			continue
		}

		switch {
		case strings.HasPrefix(line, "user-agent:"):
			ua := strings.TrimSpace(line[len("user-agent:"):])
			// start new rule if UA differs
			if current == nil || current.UserAgent != ua {
				r := RobotsRule{UserAgent: ua}
				rules = append(rules, r)
				current = &rules[len(rules)-1]
			}
		case strings.HasPrefix(line, "allow:"):
			if current != nil {
				path := strings.TrimSpace(line[len("allow:"):])
				if path != "" {
					current.AllowedPaths = append(current.AllowedPaths, path)
				}
			}
		case strings.HasPrefix(line, "disallow:"):
			if current != nil {
				path := strings.TrimSpace(line[len("disallow:"):])
				if path != "" {
					current.DisallowedPaths = append(current.DisallowedPaths, path)
				}
			}
		default:
			// if no current rule, assume wildcard
			if current == nil {
				r := RobotsRule{UserAgent: "*"}
				rules = append(rules, r)
				current = &rules[len(rules)-1]
			}
		}
	}
	return rules
}

type agentPaths struct {
	Allowed    []string
	Disallowed []string
}

func getAllowedPaths(rules []RobotsRule, userAgents []string, base string) map[string]agentPaths {
	res := make(map[string]agentPaths, len(userAgents))
	uaLower := make([]string, len(userAgents))
	for i, ua := range userAgents {
		uaLower[i] = strings.ToLower(ua)
		res[uaLower[i]] = agentPaths{}
	}

	// collect wildcard
	var wildcard []RobotsRule
	for _, r := range rules {
		if r.UserAgent == "*" {
			wildcard = append(wildcard, r)
		}
	}

	for _, ua := range uaLower {
		// find specific
		var applicable []RobotsRule
		for _, r := range rules {
			if r.UserAgent == ua {
				applicable = append(applicable, r)
			}
		}
		if len(applicable) == 0 {
			applicable = wildcard
		}
		if len(applicable) == 0 {
			// no rules => everything allowed
			res[ua] = agentPaths{Allowed: []string{base + "/"}}
			continue
		}

		ap := agentPaths{}
		for _, r := range applicable {
			if len(r.DisallowedPaths) == 0 {
				ap.Allowed = []string{base + "/"}
				break
			}
			for _, p := range r.AllowedPaths {
				p = strings.ReplaceAll(p, "*", "")
				if !strings.HasPrefix(p, "/") {
					p = "/" + p
				}
				ap.Allowed = append(ap.Allowed, base+p)
			}
			for _, p := range r.DisallowedPaths {
				p = strings.ReplaceAll(p, "*", "")
				if !strings.HasPrefix(p, "/") {
					p = "/" + p
				}
				ap.Disallowed = append(ap.Disallowed, base+p)
			}
		}
		if len(ap.Allowed) == 0 {
			root := base + "/"
			if !contains(ap.Disallowed, root) {
				ap.Allowed = append(ap.Allowed, root)
			}
		}
		res[ua] = ap
	}

	// filter wildcard disallowed from others
	if w, ok := res["*"]; ok && len(userAgents) > 1 {
		wset := make(map[string]struct{})
		for _, d := range w.Disallowed {
			wset[d] = struct{}{}
		}
		for _, ua := range uaLower {
			if ua == "*" {
				continue
			}
			ap := res[ua]
			filtered := ap.Disallowed[:0]
			for _, d := range ap.Disallowed {
				if _, blocked := wset[d]; !blocked || contains(ap.Allowed, d) {
					filtered = append(filtered, d)
				}
			}
			ap.Disallowed = filtered
			res[ua] = ap
		}
	}

	return res
}

func contains(slice []string, s string) bool {
	for _, x := range slice {
		if x == s {
			return true
		}
	}
	return false
}